src/util/timer.cpp

```diff
@@ -1,246 +1,113 @@
 #include "timer.hpp"
 
-#include <thread>
-#include <sstream>
 #include <mutex>
-
-#include "util/exception.hpp"
-
-#define ENABLED OTTO_ENABLE_TIMERS
+#include <thread>
+#include <memory>
+#include <fstream>
+#include <plog/Log.h>
+#include "core/globals.hpp"
 
 namespace otto::util::timer {
+#if OTTO_ENABLE_TIMERS
+  struct ThreadTimer {
+    ThreadTimer(std::string id) : id(id), active_timer(nullptr) {}
 
-// otto::util::timer::Timer ///////////////////////////////////////////////////
+    std::string id;
+    std::vector<TimerData> history;
+    ScopedTimer * active_timer;
 
-Timer::Timer(const timer_id& id)
-#if ENABLED
-  : id(id)
-#endif
-{}
-
-Timer::Timer(timer_id&& id)
-#if ENABLED
-  : id(std::move(id))
-#endif
-{}
+    nlohmann::json to_json() const
+    {
+      auto obj = nlohmann::json::object();
 
-void Timer::start()
-{
-#if ENABLED
-  start_time = time_point::clock::now();
-  running    = true;
-#endif
-}
+      if (!this->history.empty()) {
+        auto history = nlohmann::json::array();
 
-void Timer::stop()
-{
-#if ENABLED
-  data.push_back(time_point::clock::now() - start_time);
-  running = false;
-#endif
-}
+        for (auto const& timer_data : this->history) {
+          history.push_back(timer_data.to_json());
+        }
 
-void Timer::tick()
-{
-#if ENABLED
-  auto now = time_point::clock::now();
-  if (running) {
-    data.push_back(now - start_time);
-  } else {
-    start_time = now;
-    running    = true;
-  }
-#endif
-}
+        obj["history"] = history;
+      }
 
-nlohmann::json Timer::serialize() const
-{
-#if ENABLED
-  auto arr = nlohmann::json::array();
-  for (auto&& dur : data) {
-    arr.push_back(dur.count());
-  }
-  auto json_chldrn = nlohmann::json::object();
-  for (auto&& c : children) {
-    json_chldrn[c.id] = c.serialize();
-  }
-  return {{"data", arr}, {"children", json_chldrn}};
-#else
-  return {};
-#endif
-}
+      return obj;
+    }  
+  };
 
-// otto::util::timer::ScopedTimer /////////////////////////////////////////////
+  static std::mutex mutex;
+  static std::map<std::string, ThreadTimer> thread_timers;
+
+  static std::string get_thread_id() {
+    static thread_local std::string thread_id;
+    if (thread_id.empty()) {
+      std::ostringstream s;
+      s << "Thread " << std::this_thread::get_id();
+      thread_id = s.str();
+    }
 
-  ScopedTimer::ScopedTimer(Timer& timer) noexcept
-#if ENABLED
-    : timer (&timer)
-#endif
-  {}
+    return thread_id;
+  }
 
-  ScopedTimer::ScopedTimer(Timer* timer) noexcept
-#if ENABLED
-    : timer (timer)
-#endif
-  {}
+  static ThreadTimer& get_thread_timer() {
+    std::lock_guard lock(mutex);
+    auto thread_id = get_thread_id();
+    auto [iter, created] = thread_timers.emplace(thread_id, thread_id);
```

Comment:
could we store the actual `ThreadTimer` in a `static thread_local` here, instead of in `get_thread_id()`? that should be quite good for performance

Reply:
The problem with that is serialization.  We need a single thread to have access to the timing data for *all* threads.  If the ThreadTimer were stored in a `thread_local`, there'd be no way to access it.

Also, performance *shouldn't* be an issue here, as std::map lookups are quite fast, and timers will never be enabled in non-developer builds.

Reply:
but we could cache a refernce to it, couldnt we?

Reply:
Done.